FV3/atmos_cubed_sphere/tools/coarse_grained_diagnostics.F90

```diff
@@ -1,24 +1,232 @@
+  integer :: DIAG_SIZE = 512
+  type(coarse_diag_type), dimension(512) :: coarse_diagnostics
+
+  ! Reduction methods
+  character(len=11) :: AREA_WEIGHTED = 'area_weighted'
+  character(len=11) :: MASS_WEIGHTED = 'mass_weighted'
+  character(len=5) :: pressure_level_label
 
 contains
-  
-  subroutine fv_coarse_diag_init(Time, id_pfull, id_phalf, coarse_graining)
+
+  subroutine populate_coarse_diag_type(Atm, coarse_diagnostics)
+    type(fv_atmos_type), intent(in), target :: Atm(:)
+    type(coarse_diag_type), intent(out) :: coarse_diagnostics(:)
+
+    integer :: is, ie, js, je, npz, n_tracers, n_prognostic, t, p, n_pressure_levels
+    integer :: index = 1
+    character(len=128) :: tracer_name
+    character(len=256) :: tracer_long_name, tracer_units
+    character(len=8) :: DYNAMICS = 'dynamics'
+    integer :: pressure_levels(31)
+
+    n_pressure_levels = 31
+    pressure_levels = (/1,2,3,5,7,10,20,30,50,70,100,150,200,250,300,350,400,450,500,550,600,650,700,750,800,850,900,925,950,975,1000/)
+    npz = Atm(tile_count)%npz
+    n_prognostic = size(Atm(tile_count)%q, 4)
+    n_tracers = Atm(tile_count)%ncnst
+    call get_fine_array_bounds(is, ie, js, je)
+
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'omega_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained pressure velocity'
+    coarse_diagnostics(index)%units = 'Pa/s'
+    coarse_diagnostics(index)%reduction_method = AREA_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%omga(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'ucomp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained zonal wind'
+    coarse_diagnostics(index)%units = 'm/s'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%ua(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'vcomp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained meridional wind'
+    coarse_diagnostics(index)%units = 'm/s'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%va(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'temp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained temperature'
+    coarse_diagnostics(index)%units = 'K'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%pt(is:ie,js:je,1:npz)
+
+    do t = 1, n_tracers
```

Comment:
I like the use of 't'. The 'q' in the dynamics part of the code is super confusing, because they aren't all mixing ratio variables